### DISCUSSION TITLE
New Synesthesia Modes and the problem of coloring objects

### DISCUSSION THREAD

**Discussion by Paladynee** [ppy/osu#24342] [Ideas]
## The problem of coloring objects

The problem is that there are an infinite amount of ways to communicate timing distance using the colors of hitobjects, and I've proposed some solutions to achieve this.

# Enhancing Synesthesia with new modes

### My proposals are:

1. Mode 2 (`Next Object`)
2. Mode 4 (`Absolute Next Object`, `Absolute Previous Object`)
3. Mode 5 (`Smart Next`, `Smart Previous`)

### Here's a short clip of `Smart Next` coloring of [DeviousPanda's II-L - SPUTNIK-3](https://osu.ppy.sh/beatmapsets/1312076#osu/2719326): (where orange = 3/4, dark purple = 2/3, red = 1/2)

https://github.com/ppy/osu/assets/62181440/0b98efd8-4430-4e87-af3f-ab6e4109135d

Note how `Smart Next` _clumps_ the (4th, 5th, 6th from start) and (7th, 8th, 9th from start) together, but since they have the same signature, they are colored the same (all purple). However there is a 1/2 gap in between the _clumps_ themselves. `Smart` can't communicate how much time is between each _clump_. This is the flaw of `Smart`. But `Smart` communicates how every object of the _clump_ interacts with the each other very well.

Since mappers tend to push _clumps_ away from eachother when mapping, `Smart Next/Previous` behaves well.

### Here's a short clip of `Previous Object` coloring of the same map:

https://github.com/ppy/osu/assets/62181440/8360745e-ad0e-4cca-ac9d-2f30cece032a

Note how the coloring looks slightly random, this is caused by the visual spacings and the timing distance information clashing.
But if someone gets used to it, unlike `Smart Previous`, it conveys every single timing distance information without loss.

# List of colors I've used in this post to denote timing distance information

![image](https://github.com/ppy/osu/assets/62181440/f9fa6aff-770d-440d-a7cb-9824118a3992)

# Modes

### Chart of all Modes and their colorings

![image](https://github.com/ppy/osu/assets/62181440/211dd2e0-7444-4a48-bf9f-90a0cf407c12)

You might think that `Next Object` and `Smart Previous` are the same thing, but for this pattern, for example, they are not:

![image](https://github.com/ppy/osu/assets/62181440/3fcbfd73-d916-41a4-acd3-68050243877e)

## Mode 1: Current Behavior

Synesthesia colors the hitobject according to the editor beat snap divisor colours.

## Mode 2: Next Object

Synesthesia colors the hitobject based on the time between the next hitobject and itself.

## Mode 3: #23985's Previous Object

Synesthesia colors the hitobject based on the time between the previous hitobject and itself.

# Mode 4: Absolute Next/Previous Coloring

Synesthesia colors hitobject based on how much time is between the next/previous hitobject (either one is exactly just the other one, but shifted by 1 hitobject), following a gradient of colors (red: most amount of time between 2 objects, blue: least amount of time between 2 objects).

1/1 appears red, 1/2 appears purple, 1/3 appears a dark purple close to blue, 1/4 appears blue etc.

Here's a visualization:

![image](https://github.com/ppy/osu/assets/62181440/3b842bb2-ed2c-4fc0-9d39-bfe5f9f31fd1)

And here's the gradient where the left side represents a slower rhythm.

![image](https://github.com/ppy/osu/assets/62181440/4b190930-0022-4a7d-baaf-d1cc2d4e341c)

# Mode 5: Smart Next / Previous

Synesthesia groups hitobjects which have the same timing distance between themselves, named _clumps_ to achieve a slightly lossy but a very readable coloring.

## Smart Coloring Step 1: _Clumping_

Defining a _clump_: Group of sequential hitobjects which all have the same timing distance between them.

### How _clumping_ works:

Sequentially analyze hitobjects. For each circle, save the information of **delta time:** timing distance between the current and the next hitobject.

Whenever you encounter a different **delta time** value:
For `Smart Next`: The hitobjects after `the last clump` and before `the hitobject with the different delta time value` including `itself` is a new _clump_.

![image](https://github.com/ppy/osu/assets/62181440/7585e96e-94a7-4f28-b71f-384c138c43ec)

For `Smart Previous`: The hitobjects after `the last clump` and before `the hitobject with the different delta time value` is a new _clump_.

![image](https://github.com/ppy/osu/assets/62181440/b0b46fb0-78a4-4125-8e56-ff36c1426287)

## Difference of `Smart Next` and `Smart Previous` when _clumping_

Take this pattern as an example.

![image](https://github.com/ppy/osu/assets/62181440/90acd6a7-9163-4ac4-b827-0fb702a85ab6)

The divisors between 1-2, 2-3, 3-4... etc are as follows:
one 3/4 two 3/4 three 2/3 four 2/3 five 2/3 six

`Smart Next` will _clump_ (1, 2, 3) and (4, 5, 6) together while `Smart Previous` _clumps_ (1, 2) and (3, 4, 5, 6) together.

### This is to compensate for players who read the same pattern in 2 different ways:

**Reading 1** (player will benefit from using `Smart Next`):

![image](https://github.com/ppy/osu/assets/62181440/5f64d307-88d6-4d75-a98e-c274c9eae5cb)

**Reading 2** (player will benefit from using `Smart Previous`):

![image](https://github.com/ppy/osu/assets/62181440/158a17de-ad28-46ad-bd07-90ad87b9b9af)

## Here's another example:

![image](https://github.com/ppy/osu/assets/62181440/0666cbeb-d64e-4359-b977-c6268b720eac)

**Do you read it like this...** (you will benefit from using `Smart Next`):

![image](https://github.com/ppy/osu/assets/62181440/996958e8-fa75-47f0-83d5-83e6d58833b8)

**... or like this?:** (you will benefit from using `Smart Previous`):

![image](https://github.com/ppy/osu/assets/62181440/ed4f826d-f2bd-4223-8605-2a1993eaa7c7)

After we successfully _clump_ our objects, we start coloring them.

## Smart Coloring Step 2: Coloring

For every _clump_ that has more than 1 object, we color every object the same color, according to the timing distance's respective color. (every object of a _clump_ will have the same timing distance)

![image](https://github.com/ppy/osu/assets/62181440/ade123ef-89d1-46b8-b420-f84144649117)

For example: if we're coloring a _clump_ of 3 objects of 2/3 beat snap, all 3 objects are colored the color for 2/3.

## After our coloring, we're just left with _clumps_ of singular objects...

...where they can have the same or different timing distances for the next and previous objects.

Here's the 3 possible states of a _clump_ consisting of a singular object:

The timing distance from the previous and to the next is the same.
![image](https://github.com/ppy/osu/assets/62181440/fd92ce3d-cc90-49a9-9b58-22acea50f392)

The timing distance to the next hitobject is bigger.
![image](https://github.com/ppy/osu/assets/62181440/3ee16fed-c2b6-4440-adcd-302ff3642873)

The timing distance from the previous hitobject is bigger.
![image](https://github.com/ppy/osu/assets/62181440/87e80761-e68e-4589-8de1-f66bbbf00d4f)

How we color this object is dependant on whether we use `Smart Next` or `Smart Previous` coloring.

For `Smart Next`, we color it based on the amount of time between the single object and the next object.
For `Smart Previous`, we color it based on the amount of time between the previous object and the single object.

# Regarding sliderends

When _clumping_ and coloring, sliderends should be taken into consideration as distinct hitobjects, for example:

For example, this pattern has been colored using the `Previous Object` coloring.

![image](https://github.com/ppy/osu/assets/62181440/a49a3326-55bf-4ac0-8825-bbafb4aa0005)

Note how we didn't color the last object the color for 2 beats of timing distance, we colored it the color of 1 beat of timing distance.

And the same pattern with `Next Object`

![image](https://github.com/ppy/osu/assets/62181440/4427cc28-60c1-432a-b325-aa1ee5c94d28)

Note how the sliderhead denotes the length of the slider when using `Next Object`, instead of the timing distance between the sliderhead and the next hitcircle.

**Comment by monochrome22** [comment]
Similar to https://github.com/ppy/osu/discussions/23985.

By the way, grouping everything in the same beat snap is suboptimal, because you wouldn't be able to determine the timings between groups, that's why I suggested showing timing of the next object instead. It's slightly more complicated that way, but the functionality is there, and will be more useful once the player gets used to it.

**Reply by Paladynee** [reply]
> because you wouldn't be able to determine the timings between groups

this is false. the timings between groups are determined by the last object of the previous group.
if it follows a different timing signature between the 2 groups of objects, simply the last object of the previous group is colored differently.
this coloring issue will persist even if you change which object you're coloring the objects based on (next object / previous object).

In my opinion, both could be implemented, under mode names `Beat Snap` (current behavior), `Previous object`, `Next object`.

**Reply by monochrome22** [reply]
Could you draw an image of what your idea would look like using the same timings as [my example with the purple and white circles](https://github.com/ppy/osu/discussions/23985)?

**Reply by Paladynee** [reply]
I was doing a slightly different thing, but I think you will get the idea.
![image](https://github.com/ppy/osu/assets/62181440/dc873431-7a47-4183-8ebd-cd6d197142eb)

**Reply by Paladynee** [reply]
As for your request, heres how every mode looks:
Beat snap:
![image](https://github.com/ppy/osu/assets/62181440/e2230515-7e9a-497a-8f6b-38c763672164)
Next object:
![image](https://github.com/ppy/osu/assets/62181440/6e4f5265-3457-4694-873b-70c9db19841c)
Previous object:
![image](https://github.com/ppy/osu/assets/62181440/b6edf8ef-637e-4fc7-8481-172f85e387e2)

Gray: 1/1
Light blue: 1/4
Red: 1/2
Black: Undefined, since there is no next/previous object. Could inherit the color of the objects before/after them

**Reply by monochrome22** [reply]
And which idea are you proposing in this thread, the "Next Object" one or "Prev. Object"? Or both?

**Reply by monochrome22** [reply]
Looks like your proposal is to use what you call "Next Object".

You should not incorporate my proposal into your thread. They are two separate suggestions and should be kept in separate discussion threads.

Personally I think using what you call "Previous object" is better. With "Next Object", you would have to memorize what the color of the circle was that you just hit to know what the timing of the next circle is, while with "Previous Object", you can just see what the timing is without having to memorize anything.

**Reply by Paladynee** [reply]
I did not incorporate your proposal in my thread. Your thread was flawed, in the way of speaking about next objects and showing the previous object implementation. Both of our proposals are the next object coloring, but yours showed previous object as an example.
I just mentioned your post and made another chart for it, visualizing what you have visualized, but not talked about before.

About keeping the threads apart, they are about the same mod, same subject and just doing the same thing 2 different ways. 2 things that are this close to being the same thing should be placed under a single post, in my opinion. Your implementation of `Previous Object` coloring is just the `Next Object` color but shifted 1 to the right.

If you want the credit for the `next/previous object` modes, have it. My other proposal is the `smart` mode.

**Reply by monochrome22** [reply]
>I did not incorporate your proposal in my thread.

By incorporating my proposal, I mean linking my discussion and writing a duplicate of my proposal.

> Your thread was flawed, in the way of speaking about next objects and showing the previous object implementation.

My thread is not flawed, you're arguing semantics.

>Your implementation of `Previous Object` coloring is just the `Next Object` color but shifted 1 to the right.

Your "Previous object" proposal is the exact same as my proposal, there's no difference.

**Reply by Paladynee** [reply]
> My thread is not flawed, you're arguing semantics.

> Mod setting for Synesthesia to base the color on the timing of the next object

Explain why object 4 is colored purple?

![colors5](https://github.com/ppy/osu/assets/122969285/09a75a27-5730-4a96-bdd6-012e0ebe1022)

You are literally speaking and showing different things.


> Your "Previous object" proposal is the exact same as my proposal, there's no difference.

The only difference between the `next object` and the `previous object` colorings are just shifting by 1 object. You can clearly see this in the chart I prepared.

> ...and writing a duplicate of my proposal.

I did not write a duplicate of your proposal. When I saw your proposal, I thought to myself "hey, these are almost the same thing!" and mentioned it in my post. I don't see any duplicates here.

**Comment by monochrome22** [comment]
The `smart` idea doesn't work if the timings inbetween are irregular.

![Irregular timings](https://github.com/ppy/osu/assets/122969285/81d207df-9f8a-440d-a810-2d11145e0ac3)

In a case like this, if you base the color of (1) on (12345), then you won't know the timing from (1) to (123). If you base the color of (1) on the timing from (1) to (123), then you won't know the timing from (12345) to (1).

I pointed out this issue in my other comment.
>grouping everything in the same beat snap is suboptimal, because you wouldn't be able to determine the timings between groups

**Reply by Paladynee** [reply]
> This is not always the case though, if (7,8,9) were shifted to the right 2 * 1/4 snaps, we would have 2 ways to color object 6: next (gray) or previous (red).

**Reply by monochrome22** [reply]
Yeah, so the idea doesn't work. Whichever way you color it, you won't know the timing from 6 to the other section.

**Reply by Paladynee** [reply]
`Next/Previous Object` also fails here. There is no way to represent 3/4 of a beat in `next/previous object` coloring (what you and i both proposed) without using different colours for each numerator in a beat snap divisor (1 color for 3/4, 2 colors for 2/6 and 4/6 and so on)

**Reply by monochrome22** [reply]
There are many colors that can be added, so they can be made to work, only need 16 colors to make any snap divisor work, anything beyond 1/1 would be same color as 1/1. But for the `smart` idea, there's nothing you can do to make it work. 

But if you want a more normal example (of where next/previous would easily work and `smart` doesn't), then here's one:
![Irregular Timings 2](https://github.com/ppy/osu/assets/122969285/39e03252-867e-4420-91eb-abe891f03a1e)

**Reply by Paladynee** [reply]
> there's nothing you can do to make it work.

You are attacking the idea. I'm unsure if I follow. There is an infinite amount of ways you can communicate beat snap information using the color of an object. 

> But if you want a more normal example (of where next/previous would easily work and `smart` doesn't), then here's one

> This is not always the case though, if (7,8,9) were shifted to the right 2 * 1/4 snaps, we would have 2 ways to color object 6: next (gray) or previous (red).

Smart works in the form of clumping objects together. In your example, the first 5 objects are a clump, object 6 itself is a clump, objects 7,8,9 are a clump. There are 2 correct ways to color object 6. Not "nothing".

**Reply by monochrome22** [reply]
If you are going to color the middle circle so that you know the timing from here to here: 

![here to here 1](https://github.com/ppy/osu/assets/122969285/3beca109-d71a-45b3-8376-4bed8cfd9334)

Then you won't know the timing from here to here:

![here to here 2](https://github.com/ppy/osu/assets/122969285/d626e476-b74c-4aee-bfef-7f2dbde66569)

---

If you are going to color the middle circle so that you know the timing from here to here: 

![here to here 2](https://github.com/ppy/osu/assets/122969285/d626e476-b74c-4aee-bfef-7f2dbde66569)

Then you won't know the timing from here to here:

![here to here 1](https://github.com/ppy/osu/assets/122969285/3beca109-d71a-45b3-8376-4bed8cfd9334)

There's no way to make it work. I'm not attacking it, just pointing out the flaw.

**Reply by Paladynee** [reply]
You can't communicate 2 different divisor infos with a single color. I think that's what you're trying to say. Then Smart would also be divided into 2 parts `Smart Next` and `Smart Previous`, which `Smart Next` will incorporate image 1&2, and `Smart Previous` would incorporate image 3&4.

> There are many colors that can be added, so they can be made to work, only need 16 colors to make any snap divisor work...

The issue is not running out of colours, the issue is having the player distinguish between 2 colours and not confuse them together (or using the least amount of colours).

**Reply by monochrome22** [reply]
>Then Smart would also be divided into 2 parts Smart Next and Smart Previous, which Smart Next will incorporate image 1&2, and Smart Previous would incorporate image 3&4.

What would that look like?

**Reply by Paladynee** [reply]
`Smart Next` would color object 6 gray (time difference between object 6 and 7 is 1/1).

![image](https://github.com/ppy/osu/assets/62181440/c030b997-88f2-4eef-84d4-550d81ab97dc)

~~`Smart Previous` would color object 6 red (time difference between object 5 and 6 is 1/2).~~ `Smart Previous` no longer clumps objects together like this.

![image](https://github.com/ppy/osu/assets/62181440/a1c46e89-ca8c-46c1-83a0-050702e7e0d3)

If the time difference between object 6 and 7 were 3/4, we would need another color for `Smart Next` when it's coloring object 6...

![image](https://github.com/ppy/osu/assets/62181440/98d24b5f-bb98-4c29-a99e-928ae3fd8a5d)

~~...but `Smart Previous` wouldn't need a new colour because 7, 8 and 9 are a different clump.~~ `Smart Previous` no longer clumps objects together like this.

![image](https://github.com/ppy/osu/assets/62181440/7044442b-9ca1-49b7-96d4-293e38737130)

...adding to the list of colours that the player might confuse together. This issue persists even if we use the regular `Next/Previous Object` colorings, this applies to all of them but `Absolute Next/Previous` and `Beat Snap`.

**Reply by monochrome22** [reply]
I see, you meant dividing the `smart` idea into 2 separate mod options. So players would choose between not knowing timing from 6 to 7 or not knowing timing from 5 to 6. 
I don't know if this would get picked over just using next/previous object, but will see what devs think.

By the way, lack of colors isn't really an issue. As long as 1/1 1/2 1/3 1/4 1/6 work well, then the mod is good. Things like 3/4 are rare, so it's fine even if their color isn't as clear.

**Reply by Paladynee** [reply]
Here's a short clip of `Smart Next Object` coloring of [DeviousPanda's II-L - SPUTNIK-3 [Beyond OWC]](https://osu.ppy.sh/beatmapsets/1312076#osu/2719326):

https://github.com/ppy/osu/assets/62181440/0b98efd8-4430-4e87-af3f-ab6e4109135d

Note how `Smart Next Object` clumps the (4,5,6) and (7,8,9) but since they have the same signature, they are colored the same (all purple). But there is a slightly smaller gap in between the clumps themselves. This is the flaw of `Smart`. It can't communicate how much time is between each clump but communicates how a clump behaves very well.

Since mappers tend to push clumps away from eachother when mapping, `Smart Next/Previous Object` behaves well.

Here's a short clip of `Next Object` coloring of the same map:

https://github.com/ppy/osu/assets/62181440/bb305964-179e-49b3-9651-4080cea018fb

Note how the coloring looks very random, this is caused by the visual spacings and the timing information clashing.
But if someone gets used to it, unlike `Smart Next Object`, it conveys every single beat snap information without loss.

**Comment by monochrome22** [comment]
>Here's a short clip of Previous Object coloring of the same map:
>Note how the coloring looks slightly random, this is caused by the visual spacings and the timing information clashing.

I don't know if it's preference, but `Previous Object coloring` looks perfect to me. At least to me, the `smart next object` preview is the one that looks confusing.
Could you show a preview of `smart previous object` ?

And as I thought, `Next Object` also looks confusing, because while hitting the circle, I have no idea what the timing is. I would have to memorize the color of the previous circle that I hit to know the timing of the circle that I'm currently trying to hit.

**Reply by Paladynee** [reply]
In my opinion, `Smart Next Object` looks perfect since thats exactly how I read maps.

Here's your request of `Smart Previous Object` coloring of the same map: 

https://github.com/ppy/osu/assets/62181440/7b604aa4-eb96-401a-9848-174f45e37c92

**Reply by Paladynee** [reply]
~~I might've confused how `Smart Previous Object` 's clumping works. I'll rethink this through.~~

Nope, it works as intended.

**Reply by monochrome22** [reply]
Ok, `smart previous object` also looks confusing to me.

> In my opinion, `Smart Next Object` looks perfect since thats exactly how I read maps.

That's interesting. It might be that the best option is different for each player. That's something to keep in mind for whoever is interested in implementing additional settings for Synesthesia.

**Comment by tutacat** [comment]
*SinAesthesia
